src/main/java/uk/gov/companieshouse/controller/DissolutionController.java

```diff
@@ -70,7 +84,31 @@ public DissolutionCreateResponse submitDissolutionRequest(
             throw new ConflictException("Dissolution already exists");
         }
 
-        return dissolutionService.create(body, companyNumber, userId, request.getRemoteAddr(), getEmail(authorisedUser));
+        final CompanyProfileApi companyProfileApi;
+        final String companyName;
+
+        try {
+            companyProfileApi = companyService.getCompanyProfile(companyNumber);
+
+            if (companyProfileApi == null) {
+                throw new CompanyNotFoundException();
+            }
+
+            if (
+                    !companyService.isCompanyClosable(companyProfileApi) ||
+                    !companyOfficerService.hasEnoughOfficersSelected(companyNumber)
+            ) {
+                throw new CompanyNotClosableException();
+            }
+        } catch (ServiceException e) {
+            throw new RuntimeException();
+        }
+
+        companyName = companyProfileApi.getCompanyName();
+
+        logger.debug("[POST] Submitting dissolution request for company number {}", companyNumber);
+
+        return dissolutionService.create(body, companyNumber, companyName, userId, request.getRemoteAddr(), getEmail(authorisedUser));
```

Comment:
I believe this can be rewritten to:
```
logger.debug("[POST] Submitting dissolution request for company number {}", companyNumber);

final CompanyProfileApi companyProfile = Optional
    .ofNullable(companyService.getCompanyProfile(companyNumber))
    .orElseThrow(NotFoundException::new);

dissolutionValidator
     .checkBusinessRules(companyProfile)
     .ifPresent(error -> throw new BadRequestException(error))

return dissolutionService.create(body, companyProfile, userId, request.getRemoteAddr(), getEmail(authorisedUser));
```

where the DissolutionValidator is a new service who is responsible for checking business rules and returns an error if one is violated, i.e.:
```
if (company is not closable) {
  return Optional.of("Company must be ...")
}

if (majority of officers are not selected) {
  return Optional.of("A majority of directors must be selected")
}

return Optional.empty()
```

This gives more granular error messaging and also makes the controller simpler. For example we currently have no way of identifying why a BadRequest was thrown. Which of the business rules failed?

Reply:
Structure wise where would we keep the Validator?

Reply:
probs in the services package